byterun/floats.c

```diff
@@ -226,7 +226,14 @@ static int caml_float_of_hex(const char * s, double * res)
       if (*s == 0) return -1;   /* nothing after exponent mark */
       e = strtol(s, &p, 10);
       if (*p != 0) return -1;   /* ill-formed exponent */
-      if (e < INT_MIN || e > INT_MAX) return -1; /* unreasonable exponent */
+      if (e <= INT_MIN) {
+        *res = 0.;
+        return 0;
+      }
+      else if (e >= INT_MAX) {
+        *res = m == 0 ? 0. : HUGE_VAL;
+        return 0;
+      }
```

Comment:
This change is questionable.

[MPR](https://caml.inria.fr/mantis/view.php?id=7690) states:
> The test "if (e < INT_MIN || e > INT_MAX) return -1;" is ineffective on Win64 since int and long have the same size there: this results in some difference in behavior between win64 and linux.

What is the difference? Previously, the conversion failed when `e` did not fit into an `int`, regardless of whether `int` and `long` have the same width (as if we used an imaginary `strtoi` function); but now the behaviour depends on the size of `long`.

Also, `INT_MIN` and `INT_MAX` are now special cases. Why were `<` and `>` changed into `<=` and `>=`?

Reply:
I think I agree with the change.  There are two cases where the exponent overflows:
* `strtol` overflows because the exponent is not representable as a `long`, in which case it returns `LONG_MAX` or `LONG_MIN`.
* the `long` returned by `strtol` cannot be represented as an `int`.

The old test `e < INT_MIN || e > INT_MAX` catches both issues at once but only if `LONG_MIN < INT_MIN` and `LONG_MAX > INT_MAX`, which is not the case under Windows.

The new tests `e <= INT_MIN` and `e >= INT_MAX` catches both issues even if `LONG_MIN == INT_MIN` and `LONG_MAX == INT_MAX` as in Windows.

Treating `e == INT_MIN` as an underflow and `e == INT_MAX` as an overflow is OK floating-point wise.

Reply:
Got it: `strtol` overflow was not detected on Windows. It makes sense now. Thank you!

The change looks correct as it is, but perhaps it'd be better to check `errno`, if only for clarity?
```c
errno = 0;
e = strtol(s, &p, 10);
/* out of range of type long? */
if (errno == ERANGE && (e == LONG_MIN || e == LONG_MAX)) return -1;
/* out of range of type int? */
if (e < INT_MIN || e > INT_MAX) return -1;
exp = (int)e;
```

Reply:
This check in the code above seems redundant: `e == LONG_MIN || e == LONG_MAX`. I took it from [CERT C](https://wiki.sei.cmu.edu/confluence/display/c/ERR34-C.+Detect+errors+when+converting+a+string+to+a+number).

Reply:
Yes, we could follow the pattern of tests from the CERT example, because it is a lot clearer!  But note that `return -1` is no longer the correct action.  I'm torn about efficiency: string->float conversion may be time-critical and  the CERT code performs more tests than @oandrieu's code.  On the other hand calling `strtod` and `ldexp` isn't cheap already.

Reply:
An explicit `#ifdef` could be even better, for both clarity and performance:
```c
#if INT_MAX == LONG_MAX
      if (errno == ERANGE) return -1;
#else
      if (e < INT_MIN || e > INT_MAX) return -1;
#endif
```

This should make it more obvious to the reader that `int` can have the same width as `long`, and that it affects error-checking.

Reply:
But is there a need to make this function fail when the exponent overflows ? I don't see the point. What's wrong with returning 0 / ∞ ?

Personally, I prefer a version that doesn't use the preprocessor nor a global value.
I can add comments to explain that the test is laid out this way because of Win64.

Reply:
> But is there a need to make this function fail when the exponent overflows ? I don't see the point. What's wrong with returning 0 / ∞ ?

Nothing wrong. I think it's a slight improvement, and was mostly confused by the error detection logic, which I found totally unobvious at first sight.


> I can add comments to explain that the test is laid out this way because of Win64.

Another fine option, yes.

Reply:
OK I added a comment.

Reply:
The comment brings clarity. Thank you!